### PR TITLE
Cleanup and add support for chunking

### PR DESCRIPTION
**Pull request by Universal-Omega**
Also does other major cleanup/modernization and some rewrites.

### CODE REVIEW

File: includes/Api/ApiViewDumps.php [hunk 1]
```diff
@@ -77,22 +77,11 @@ private function getDownloadUrl( $config, $dump ) {
 			return $dump->dumps_filename;
 		}
 
-		// If wgDataDumpDownloadUrl is configured, use that
```
Comment (paladox) [inline includes/Api/ApiViewDumps.php:80]: This should not be removed. It means we are using php which reduces capacity & slowness & possible memory issues.

Comment (Universal-Omega) [reply includes/Api/ApiViewDumps.php:80]: > This should not be removed. It means we are using php which reduces capacity & slowness & possible memory issues.

It cannot stream chunks if we use this so it has to be removed, additionally I actually notice it works a lot faster by using the internal streamer instead...

File: includes/DataDump.php [hunk 2]
```diff
@@ -11,33 +11,31 @@
...
+		static $backend = null;
+		if ( !$backend ) {
+			$dirConfig = $config->get( ConfigNames::Directory );
+			$uploadDir = $config->get( MainConfigNames::UploadDirectory );
+			$backend = new FSFileBackend( [
+				'name' => 'dumps-backend',
+				'wikiId' => WikiMap::getCurrentWikiId(),
+				'lockManager' => new NullLockManager( [] ),
+				'containerPaths' => [ 'dumps-backup' => $dirConfig ?: "{$uploadDir}/dumps" ],
+				'fileMode' => 0777,
```
Comment (BlankEclair) [inline includes/DataDump.php:33]: Please use something more sensible... like 0600 or 0660, please, for the love of-

File: includes/DataDumpPager.php [hunk 3]
```diff
@@ -2,185 +2,201 @@
...
-			$msg = $this->msg( $msg )->text();
-		}
-
-		return $headers;
 	}
 
-	public function formatValue( $name, $value ) {
-		$row = $this->mCurrentRow;
+	/** @inheritDoc */
+	public function formatValue( $name, $value ): string {
```
Comment (BlankEclair) [inline includes/DataDumpPager.php:61]: Type hint?

File: includes/DataDumpPager.php [hunk 4]
```diff
@@ -2,185 +2,201 @@
...
+					) {
 						$formatted = Html::openElement(
 							'form',
 							[
-								'action' => $link,
-								'method' => 'POST'
+								'action' => $pageTitle->getLinkURL( $query ),
+								'method' => 'POST',
 							]
 						) . $element . $token . Html::closeElement( 'form' );
```
Comment (BlankEclair) [inline includes/DataDumpPager.php:137]: `Html::rawElement()`

File: includes/DataDumpPager.php [hunk 5]
```diff
@@ -2,185 +2,201 @@
...
+		$dataDumpDisableGenerate = $this->config->get( ConfigNames::DisableGenerate );
 		if ( $dataDumpDisableGenerate ) {
 			$out = $this->getOutput();
 			$out->addHTML(
 				Html::errorBox( $this->msg( 'datadump-generated-disabled' )->escaped() )
 			);
 
 			$out->addHTML(
-				'<br />' . Linker::specialLink( 'DataDump', 'datadump-refresh' )
+				Html::closeElement( 'br' ) .
```
Comment (BlankEclair) [inline includes/DataDumpPager.php:192]:
```suggestion
				Html::element( 'br' ) .
```

File: includes/DataDumpPager.php [hunk 6]
```diff
@@ -342,25 +362,22 @@ public function onGenerate( array $params ) {
...
-
-		return true;
 	}
 
-	private function getGenerateLimit( string $type ) {
-		$config = $this->config->get( 'DataDump' );
+	private function getGenerateLimit( string $type ): bool {
+		$config = $this->config->get( ConfigNames::DataDump );
 
 		if ( isset( $config[$type]['limit'] ) && $config[$type]['limit'] ) {
```
Comment (BlankEclair) [inline includes/DataDumpPager.php:370]:
```suggestion
		if ( $config[$type]['limit'] ?? null ) {
```

File: includes/DataDumpPager.php [hunk 7]
```diff
@@ -342,25 +362,22 @@ public function onGenerate( array $params ) {
...
-				[
-					'dumps_type' => $type
-				]
-			);
+			$res = $this->getDatabase()->newSelectQueryBuilder()
+				->select( '*' )
+				->from( 'data_dump' )
+				->where( [ 'dumps_type' => $type ] )
+				->caller( __METHOD__ )
+				->fetchResultSet();
```
Comment (BlankEclair) [inline includes/DataDumpPager.php:376]: [`SelectQueryBuilder::fetchRowCount()`](https://doc.wikimedia.org/mediawiki-core/REL1_42/php/classWikimedia_1_1Rdbms_1_1SelectQueryBuilder.html#ad2f7767fae0bb5bec8482ec224246677)

Probably should try the COUNT() expression example in the documention of the function linked above, but no worries if it doesn't work.

File: includes/Jobs/DataDumpGenerateJob.php [hunk 8]
```diff
@@ -5,176 +5,360 @@
...
-		$this->setStatus( 'in-progress', $dbw, '', $fileName, __METHOD__ );
-
-		$options = [];
-		foreach ( $dataDumpConfig[$type]['generate']['options'] as $option ) {
-			$options[] = preg_replace( '/\$\{filename\}/im', $fileName, $option );
-		}
+		$options = array_map(
+			fn ( string $opt ): string => preg_replace(
+				'/\$\{filename\}/im', $fileName, $opt
+			),
```
Comment (BlankEclair) [inline includes/Jobs/DataDumpGenerateJob.php:68]: I don't think we need mode `m`, so try [`str_ireplace`](https://www.php.net/manual/en/function.str-ireplace.php) (or perhaps even `str_replace`, but that might break API)

File: includes/Jobs/DataDumpGenerateJob.php [hunk 9]
```diff
@@ -5,176 +5,360 @@
...
-				->execute()
-				->getExitCode();
-		} else {
+		return $this->setStatus(
+			status: 'failed',
+			dbw: $dbw,
+			directoryBackend: $directoryBackend,
+			fileName: $fileName,
+			fname: __METHOD__,
+			comment: 'Something went wrong',
```
Comment (BlankEclair) [inline includes/Jobs/DataDumpGenerateJob.php:105]: What about a more explanatory message with the status code, such as something like `"Something went wrong: Command exited with {$result}"`

File: includes/Specials/SpecialDataDump.php [hunk 10]
```diff
@@ -79,38 +83,84 @@ public function execute( $par ) {
...
-				'Content-Disposition: attachment; filename="' . $fileName . '"',
-			]
-		] )->isOK();
+		$directoryBackend = $backend->getContainerStoragePath( 'dumps-backup' );
+
+		if ( $backend->fileExists( [ 'src' => $directoryBackend . '/' . $fileName ] ) ) {
+			$backend->streamFile( [
+				'src' => $directoryBackend . '/' . $fileName,
+				'headers' => [
+					'Expires: ' . gmdate( 'D, d M Y H:i:s', 0 ) . ' GMT',
```
Comment (BlankEclair) [inline includes/Specials/SpecialDataDump.php:112]: I wonder if we should hardcode the timestamp here

File: includes/Specials/SpecialDataDump.php [hunk 11]
```diff
@@ -79,38 +83,84 @@ public function execute( $par ) {
...
+		string $directoryBackend,
+		FileBackend $backend
+	): void {
+		$chunkIndex = 0;
+		$chunkFileName = $fileName . '.part' . $chunkIndex;
+		$headersSent = false;
+
+		while ( $backend->fileExists( [ 'src' => $directoryBackend . '/' . $chunkFileName ] ) ) {
+			if ( !$headersSent ) {
+				header( 'Expires: ' . gmdate( 'D, d M Y H:i:s', 0 ) . ' GMT' );
```
Comment (BlankEclair) [inline includes/Specials/SpecialDataDump.php:136]: I wonder if we should hardcode the timestamp here

File: includes/Specials/SpecialDataDump.php [hunk 12]
```diff
@@ -79,38 +83,84 @@ public function execute( $par ) {
...
+		);
 
 		$out->addModuleStyles( [ 'mediawiki.special' ] );
 
 		$pager->getForm();
 		$out->addParserOutputContent( $pager->getFullOutput() );
 	}
 
-	private function doDownload( string $fileName ) {
+	private function doDownload( string $fileName ): bool {
```
Comment (BlankEclair) [inline includes/Specials/SpecialDataDump.php:101]: Return value is unused

File: maintenance/insertMissingDumps.php [hunk 13]
```diff
@@ -34,39 +36,61 @@ public function execute() {
...
 
-		$missingDumps = array_diff( array_values( $dumpFiles ), $existingDumps );
-		foreach ( $missingDumps as $dump ) {
-			$fileSize = $backend->getFileSize( [
-				'src' => "$storagePath/$dump"
-			] );
+		$chunkedFiles = [];
+		foreach ( $dumpFiles as $file ) {
+			// Group files by their base names (strip .part<number>).
+			if ( preg_match( '/^(.*)\.part\d+$/', $file, $matches ) ) {
```
Comment (BlankEclair) [inline maintenance/insertMissingDumps.php:45]:
```suggestion
			if ( preg_match( '/\.part\d+$/', $file ) ) {
```

File: includes/DataDumpPager.php (hunk 3 above)
Comment (Universal-Omega) [reply includes/DataDumpPager.php:61]: Can't type hint an override function when the source isn't as then it throws errors.

File: includes/Specials/SpecialDataDump.php (hunk 11 above)
Comment (Universal-Omega) [reply includes/Specials/SpecialDataDump.php:136]: What do you mean?

File: includes/DataDumpPager.php (hunk 7 above)
Comment (Universal-Omega) [reply includes/DataDumpPager.php]: Good idea. Done.

File: includes/Jobs/DataDumpGenerateJob.php (hunk 8 above)
Comment (Universal-Omega) [reply includes/Jobs/DataDumpGenerateJob.php:68]: I think I will just leave it for now as this currently just maintains how it is done currently.

File: includes/DataDumpPager.php (hunk 5 above)
Comment (BlankEclair) [reply includes/DataDumpPager.php:193]: MediaWiki checks for that, and doesn't write the closing element if so:
https://gerrit.wikimedia.org/r/plugins/gitiles/mediawiki/core/+/1a2ac9f421b8a6d9a5717024772015a39fa8216f/includes/Html/Html.php#193

File: maintenance/insertMissingDumps.php (hunk 13 above)
Comment (BlankEclair) [reply maintenance/insertMissingDumps.php:45]: Oh my god, I missed that part somehow

File: includes/DataDumpPager.php (hunk 5 above)
Comment (Universal-Omega) [reply includes/DataDumpPager.php:193]: Oh interesting I never knew that lol

File: maintenance/insertMissingDumps.php (hunk 13 above)
Comment (Universal-Omega) [reply maintenance/insertMissingDumps.php:45]: Then matches doesn't output what we need it to, it matches match groups so like those in parentheses first group is $matches[1]

Comment (BlankEclair) [reply maintenance/insertMissingDumps.php:45]: oh my god i am embarassing myself so much :neocat_blush_hide:

**Pull request merged**